deployments/curity/apigateway/openresty/cors.lua

```diff
@@ -0,0 +1,56 @@
+--[[
+Copyright (C) 2024 Curity AB. All rights reserved.
+
+The contents of this file are the property of Curity AB.
+You may not copy or use this file, in either source code
+or executable form, except in compliance with terms
+set by Curity AB.
+
+For further information, please contact Curity AB.
+]]
+
+local cors = {}
+
+local function array_has_value(arr, val)
+    for index, value in ipairs(arr) do
+        if value == val then
+            return true
+        end
+    end
+    return false
+end
+
+function cors.run(config)
+    local origin = ngx.req.get_headers()['origin']
+    if not (origin and array_has_value(config.origins, origin)) then
+        return
+    end
+
+    local should_exit = false
+    local vary = 'origin'
+
+    ngx.header['access-control-allow-origin'] = origin
+    ngx.header['access-control-allow-credentials'] = tostring(config.credentials)
+
+    local method = ngx.req.get_method():upper()
+    if method == 'OPTIONS' then
+        should_exit = true
+
+         ngx.header['access-control-allow-methods'] = 'GET,POST,PUT,PATCH,DELETE'
+         ngx.header['access-control-max-age'] = config.max_age
+
+         local request_headers = ngx.req.get_headers()['access-control-request-headers']
+         if request_headers then
+            ngx.header['access-control-allow-headers'] = request_headers
```

Comment:
allowing every header in the request without any validation rings an alarm bell. Since this is only for the example, don't we know which headers to expect?

Reply:
We use a default of allowing any header - but only for the trusted origin - so that the behavior is the same for web and mobile apps. Note that some libraries use this behavior also, eg the Node.js CORS library. 

From a security viewpoint:

- We use CORS only to restrict access to trusted origins. We provide only a CORS demo-level solution which customers can override. CORS is no longer implemented in the OAuth proxy itself.

There is also a developer experience viewpoint:

- It is very common when new to CORS that developers run into cryptic errors, eg they send a custom header to their API and CORS breaks with a network error and no other information. We like to avoid that in our examples.

- Or people are concerned about the performance overhead of preflights, so we use the maximum MaxAge setting to indicate that these do not have to be every request.